interfaces/apparmor/cfg.go

```diff
@@ -0,0 +1,262 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package apparmor implements integration between snappy and
+// ubuntu-core-launcher around apparmor.
+//
+// Snappy creates apparmor profiles for each application (for each snap)
+// present in the system.  Upon each execution of ubuntu-core-launcher, the
+// profile is attached to the running process. Prior to that the profile must
+// be parsed, compiled and loaded into the kernel using the support tool
+// "apparmor_parser".
+//
+// Each apparmor profile contains a simple <header><content><footer> structure.
+// The header specified an identifier that is relevant to the kernel. The
+// identifier can be either the full path of the executable or an abstract
+// identifier not related to the executable name.
+//
+// The actual profiles are stored in /var/lib/snappy/apparmor/profiles.
+// This directory is also hard-coded in ubuntu-core-launcher.
+//
+// NOTE: A systemd job (TODO: specify which) loads all snappy-specific apparmor
+// profiles into the kernel during the boot process.
+package apparmor
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/interfaces"
+	"github.com/ubuntu-core/snappy/interfaces/dbus"
+	"github.com/ubuntu-core/snappy/osutil"
+	"github.com/ubuntu-core/snappy/snap"
+)
+
+// Configurator is responsible for maintaining apparmor profiles for ubuntu-core-launcher.
+type Configurator struct {
+	pendingReload []string
+	pendingUnload []Profile
+}
+
+// ConfigureSnapSecurity creates security artefacts specific to a given snap.
+// The snap can be in developer mode to make security violations non-fatal to
+// the offending application process.
+//
+// This method only deals with files. Actual kernel changes (reloading and
+// unloading profiles) is deferred to a call to Finalize()
+func (cfg *Configurator) ConfigureSnapSecurity(repo *interfaces.Repository, snapInfo *snap.Info, developerMode bool) error {
+	// Get the snippets that apply to this snap
+	snippets, err := repo.SecuritySnippetsForSnap(snapInfo.Name, cfg.SecuritySystem())
+	if err != nil {
+		return fmt.Errorf("cannot obtain security snippets for snap %q: %s", snapInfo.Name, err)
+	}
+	// Get the files that this snap should have
+	dir, glob, content, err := cfg.DirStateForInstalledSnap(snapInfo, developerMode, snippets)
+	if err != nil {
+		return fmt.Errorf("cannot obtain expected apparmor files for snap %q: %s", snapInfo.Name, err)
+	}
+	// Ensure that files are correctly on disk
+	changed, removed, err := osutil.EnsureDirState(dir, glob, content)
+	// Record changes so that we can do clean-up in Finalize. This has to
```

Comment:
All ConfigureSnapSecurity seem to start the same? first impression is that the abstraction is not the best if it ends up with that much repetition? what's the relation between Configurators and Security Systems?